yt/yt/server/master/object_server/object_service.cpp

```diff
@@ -1582,49 +1585,53 @@ class TObjectService::TExecuteSession
         return true;
     }
 
-    bool ThrottleRequests()
+    template <EExecutionSessionSubrequestType SubrequestType, EUserWorkloadType WorkloadType>
+    bool DoThrottleRequest(int* currentSubrequestIndex, int* throttledSubrequestIndex)
     {
-        auto doThrottle = [&] (
-            int* currentSubrequestIndex,
-            int* throttledSubrequestIndex,
-            EExecutionSessionSubrequestType subrequestType,
-            EUserWorkloadType workloadType)
-        {
-            while (*throttledSubrequestIndex < *currentSubrequestIndex && *currentSubrequestIndex < TotalSubrequestCount_) {
-                auto subrequestIndex = ++(*throttledSubrequestIndex);
-                if (Subrequests_[subrequestIndex].Type != subrequestType) {
-                    continue;
-                }
+        while (*throttledSubrequestIndex < *currentSubrequestIndex && *currentSubrequestIndex < TotalSubrequestCount_) {
+            auto subrequestIndex = ++(*throttledSubrequestIndex);
+            if (Subrequests_[subrequestIndex].Type != SubrequestType) {
+                continue;
+            }
+
+            const auto& securityManager = Bootstrap_->GetSecurityManager();
 
-                const auto& securityManager = Bootstrap_->GetSecurityManager();
-                auto result = securityManager->ThrottleUser(User_.Get(), 1, workloadType);
-                if (!WaitForAndContinue(result)) {
-                    YT_LOG_DEBUG("Throttling subrequest (User: %v, SubrequestIndex: %v, SubrequestType: %v, WorkloadType: %v)",
-                        User_->GetName(),
-                        subrequestIndex,
-                        subrequestType,
-                        workloadType);
-                    return false;
+            auto throttlerFuture = securityManager->ThrottleUser(User_.Get(), 1, WorkloadType);;
+
+            if constexpr (SubrequestType == EExecutionSessionSubrequestType::LocalWrite) {
+                if (User_.Get() != securityManager->GetRootUser()) {
+                    throttlerFuture = throttlerFuture.Apply(
+                        BIND_NO_PROPAGATE([localWriteThrottlerFuture = Owner_->LocalWriteRequestThrottler_->Throttle(1)] {
+                            return localWriteThrottlerFuture;
```

Comment:
I think this merits a commentary explaining out past difficulties with the user object lifetime here.